### PR TITLE
feat: #1155: Configure Atlas to be able to reject Full mode donor import files

### PR DESCRIPTION
**Pull request by SergeyEzh**
This change is [Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)

### CODE REVIEW

File: Atlas.DonorImport.Test/Services/DonorChecker/DonorInfoCheckerTests.cs [hunk 1]
```diff
@@ -170,5 +170,18 @@ public async Task CompareDonorInfoInFileToAtlasDonorStore_WhenEmptyDonorFileExce
             await notificationSender.Received().SendAlert("Donor info checker file was present but it was empty.", $"Donors file: name-of-the-file.ext", Priority.Medium, Arg.Any<string>());
         }
 
+        [Test]
+        public async Task CompareDonorInfoInFileToAtlasDonorStore_DisposeIsCalledOnILazilyParsingDonorFile()
+        {
+            var file = DonorImportFileBuilder.NewWithoutContents.With(x => x.FileLocation, "name-of-the-file.ext");
+            var fileParserResult = Substitute.For<ILazilyParsingDonorFile>();
+            fileParserResult.ReadLazyDonorUpdates().Throws<EmptyDonorFileException>();
+            fileParser.PrepareToLazilyParseDonorUpdates(default).Returns(fileParserResult);
```
Comment (zabeen) [inline Atlas.DonorImport.Test/Services/DonorChecker/DonorInfoCheckerTests.cs:179]: This set up needs explaining - why are you intentionally throwing an exception on read of the file?

I suspect this is copy/paste error.

**Review by zabeen — CHANGES_REQUESTED** [review]
I only see one test added to DonorInfoCheckerTests, looks like we need more to cover the new setting controlling full mode import

**Review by zabeen — CHANGES_REQUESTED** [review]
I only see one test added to DonorInfoCheckerTests, looks like we need more to cover the new setting controlling full mode import

File: Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs [hunk 2]
```diff
@@ -169,5 +192,26 @@ public async Task ImportDonorFile_WhenUnexpectedException_SendsFailedResultMessa
 
             await donorImportMessageSender.Received().SendFailureMessage(file.FileLocation, ImportFailureReason.ErrorDuringImport, error);
         }
+
+        [Test]
+        public async Task ImportDonorFile_DisposeIsCalledOnILazilyParsingDonorFile()
+        {
+            var file = DonorImportFileBuilder.NewWithDefaultContents.Build();
+            var lazeFile = Substitute.For<ILazilyParsingDonorFile>();
+            lazeFile.ReadLazyDonorUpdates().Throws(new Exception());
```
Comment (zabeen) [inline Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs:201]: Again, not clear to me why exception is being thrown

**Review by SergeyEzh — COMMENTED** [review]
I added test to DonorInfoCheckerTests only because I made `ILazilyParsingDonorFile` disposable and needed to ensure that all its dependencies dispose it. Despite this I didn't make any changes in  DonorInfoChecker because I thought that #1155 is only about DonorFileImporter.

> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-:-Nne8ZzdCYn3W_1FVtdK:b-xs8mau)* status: 0 of 16 files reviewed, 7 unresolved discussions (waiting on @zabeen)

___
*[`Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs` line 201 at r1](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#gh-1445013366:-Nne63koCcfp5iYVBHIa:bulenkf) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/7f8e02886d1bd82f008b869a5ead603d4a15bf1d/Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs#L201)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

Again, not clear to me why exception is being thrown
</blockquote></details>

Explained in reply for DonorInfoCheckerTests' comment.
___
*[`Atlas.DonorImport.Test/Services/DonorChecker/DonorInfoCheckerTests.cs` line 179 at r1](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#gh-1445000530:-Nne5p_tEeG6CMa1THw3:b5v7mgl) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/7f8e02886d1bd82f008b869a5ead603d4a15bf1d/Atlas.DonorImport.Test/Services/DonorChecker/DonorInfoCheckerTests.cs#L179)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

This set up needs explaining - why are you intentionally throwing an exception on read of the file?

I suspect this is copy/paste error.
</blockquote></details>

I didn't want to build substitute for `ILazilyParsingDonorFile` further than it is necessary (which could be a bit complicated). The purpose of this test is to check if `ILazilyParsingDonorFile` was disposed after using it. So this should be enough for this. Unless, you think we need to specifically check if it disposed when exception isn't thrown.

**Review by zabeen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed, 9 unresolved discussions (waiting on @SergeyEzh)

*[a discussion](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoCFel23U8gZE6uMJS2:-NoCFel23U8gZE6uMJS3:b89wdkl) (no related file):*
Adding blocking comment for myself to update donor import and atlas integration readmes once Sergey has completed his work

___

**Review by zabeen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed, 9 unresolved discussions (waiting on @SergeyEzh)

___
*[`Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs` line 23 at r3](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoCBDaXBI6hy4uTWjJr:-NoCG28P8ikDfOH6MTmE:b-eeodjo) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f26ad3604d84c64f8dcbd93f8c6d514a2c0ff824/Atlas.DonorImport.Test/DonorImportMessageSenderTests.cs#L23)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

The focus of this test fixture is to check that messages get sent out under success/failure. The tests do not assert other behaviours like calls to the file parser and the change applier service.

We need to test a very specific requirement, which is that Full mode files are accepted/rejected based on the new app setting.

For better clarity and implicit documentation, I recommend moving the Full mode unit tests to their own fixture, so we can assert on more relevant behaviours/service calls, rather than just the one aspect of messaging.
</blockquote></details>

I also think integration tests would be invaluable - also, in their own fixture - to ensure full mode files are or are not reaching the db

**Review by zabeen — CHANGES_REQUESTED** [review]
> Reviewed 5 of 8 files at r4, 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed, 8 unresolved discussions (waiting on @SergeyEzh)

___
*[`README_Configuration.md` line 42 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoprrTACxmuxRT1XK_y:-NoprrTACxmuxRT1XK_z:b-lgc38h) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/README_Configuration.md#L42)):*
> ```Markdown
> | DONOR-IMPORT-NOTIFICATIONS-ON-DELETION-OF-INVALID-DONOR | NotificationConfiguration-NotifyOnAttemptedDeletionOfUntrackedDonor | When enabled, notifications will be sent if an import file attempts to delete a donor that was not tracked in Atlas. |
> | DONOR-IMPORT-NOTIFICATIONS-ON-SUCCESSFUL-IMPORT | NotificationConfiguration-NotifyOnSuccessfulDonorImport | When enabled, notifications will be sent for every imported donor file. May want to be disabled if donor imports are very frequent. |
> | DONOR_IMPORT_ALLOW_FULL_MODE_IMPORT | DonorImport-AllowFullModeImport | Controls whether Full mode donor import files will be accepted (true) or rejected (false) |
> ```

small edit, we have to swap underscore (`_`)  for hyphen (`-`) within `md` docs else the line doesn't wrap properly. It's very annoying, but please find/replace the symbols, thanks



_Code quote:_
```Markdown
DONOR_IMPORT_ALLOW_FULL_MODE_IMPORT 
```
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 65 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Nopz7K9Dc8MnZslUxQA:-Nopz7K9Dc8MnZslUxQB:bdup9n8) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L65)):*
> ```Smalltalk
> 
>         [Test]
>         public async Task ImportDonorFile_RejectsImportFileAndSendsNotification_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsFalse()
> ```

Small thing, our test naming pattern is `[method-name]_[condition]_[expected-outcome]`

So here, it would be: `ImportDonorFile_WhenUpdateModeIsFull_AndFullModeImportNotAllowed_FailsImportFileAndSendsAlert`



_Code quote:_
```Smalltalk
ImportDonorFile_RejectsImportFileAndSendsNotification_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsFalse()
```
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 86 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NopzU365ZktyivbDV1q:-NopzU365ZktyivbDV1r:bek9y69) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L86)):*
> ```Smalltalk
>         [TestCase(UpdateMode.Differential, true)]
>         [TestCase(UpdateMode.Differential, false)]
>         public async Task ImportDonorFile_AcceptImportFile_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsTrueOrUpdateIsNotFull(UpdateMode updateMode, bool allowFullModelImport)
> ```

I would simplify name to: `ImportDonorFile_AcceptsImportFile`

I wouldn't list out all the test cases in the name, as they're sufficiently well documented by the `[TestCase]` attributes themselves.



_Code quote:_
```Smalltalk
ImportDonorFile_AcceptImportFile_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsTrueOrUpdateIsNotFull
```
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 89 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq-Wj-0AnCmdPsEiw0:-Noq-Wj-0AnCmdPsEiw1:b-ql0cq0) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L89)):*
> ```Smalltalk
>         {
>             var file = DonorImportFileBuilder.NewWithoutContents.Build();
>             settings.AllowFullModeImport = true;
> ```

should this be set by param, `allowFullModeImport`?



_Code quote:_
```Smalltalk
settings.AllowFullModeImport = true;
```
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 90 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq-lgJ4jtOBlkfFG6g:-Noq-lgJ4jtOBlkfFG6h:bsvs1qq) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L90)):*
> ```Smalltalk
>             var file = DonorImportFileBuilder.NewWithoutContents.Build();
>             settings.AllowFullModeImport = true;
>             lazilyParsingDonorFile.ReadUpdateMode().Returns(UpdateMode.Full);
> ```

should this be set by param, `updateMode`?



_Code quote:_
```Smalltalk
(UpdateMode.Full)
```
___
*[`Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs` line 17 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq12KX3xqX1FtSTbFe:-Noq12KX3xqX1FtSTbFf:b-rn5902) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs#L17)):*
> ```Smalltalk
> {
>     [TestFixture]
>     public class RejectFullModeDonorImportTests
> ```

Many thanks for adding this fixture!

Please rename to `FullModeDonorImportTests`

and add one more test which is opposite to the existing one, where you assert donors are successfully written to the db when the setting is explicitly set to `true`.



_Code quote:_
```Smalltalk
RejectFullModeDonorImportTests
```
___
*[`Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs` line 42 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq07X89a12j9otBMEq:-Noq07X89a12j9otBMEr:b412emz) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs#L42)):*
> ```Smalltalk
> 
>         [Test]
>         public async Task ImportDonorFile_DoesntWriteToDatabase_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsFalse()
> ```

rename suggestion: `ImportDonorFile_WhenUpdateModeIsFull_AndFullModeImportIsAllowed_DoesNotWriteToDatabase`



_Code quote:_
```Smalltalk
ImportDonorFile_DoesntWriteToDatabase_WhenUpdateModeIsFullAndAllowFullModeImportSettingIsFalse
```
___
*[`terraform/core/modules/donor_import/function.tf` line 28 at r3](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoCE5NXD5ql2HDiW51N:-NoprVsGCvR4QUnPANjV:b-99a3yh) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f26ad3604d84c64f8dcbd93f8c6d514a2c0ff824/terraform/core/modules/donor_import/function.tf#L28)):*
<details><summary><i>Previously, SergeyEzh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Funny that `fmt` did not change the text alignment on this line! Not a huge deal.

**Review by SergeyEzh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: 15 of 20 files reviewed, 8 unresolved discussions (waiting on @zabeen)

___
*[`README_Configuration.md` line 42 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoprrTACxmuxRT1XK_y:-NoqDiOt2HUWQU7cmVMF:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/README_Configuration.md#L42)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

small edit, we have to swap underscore (`_`)  for hyphen (`-`) within `md` docs else the line doesn't wrap properly. It's very annoying, but please find/replace the symbols, thanks
</blockquote></details>

Done.
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 65 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Nopz7K9Dc8MnZslUxQA:-NoqDkU73neZ_EcKWTyn:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L65)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

Small thing, our test naming pattern is `[method-name]_[condition]_[expected-outcome]`

So here, it would be: `ImportDonorFile_WhenUpdateModeIsFull_AndFullModeImportNotAllowed_FailsImportFileAndSendsAlert`
</blockquote></details>

Done.
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 86 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NopzU365ZktyivbDV1q:-NoqDnMPBJALosuVrvSr:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L86)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

I would simplify name to: `ImportDonorFile_AcceptsImportFile`

I wouldn't list out all the test cases in the name, as they're sufficiently well documented by the `[TestCase]` attributes themselves.
</blockquote></details>

Done.
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 89 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq-Wj-0AnCmdPsEiw0:-NoqDpzu0SsKbSpnhncy:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L89)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

should this be set by param, `allowFullModeImport`?
</blockquote></details>

Done.
___
*[`Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs` line 90 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq-lgJ4jtOBlkfFG6g:-NoqDr93Clddxgeas3La:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test/Services/DonorFileImporterTests.cs#L90)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

should this be set by param, `updateMode`?
</blockquote></details>

Done.
___
*[`terraform/core/modules/donor_import/function.tf` line 28 at r3](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-NoCE5NXD5ql2HDiW51N:-NoqE3-D4Al7wbTign66:b-2233sx) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f26ad3604d84c64f8dcbd93f8c6d514a2c0ff824/terraform/core/modules/donor_import/function.tf#L28)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

Funny that `fmt` did not change the text alignment on this line! Not a huge deal.
</blockquote></details>

It did. It is my fault, should have rested and reviewed again before commit. Also didn't get used to `reviewable.io`
___
*[`Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs` line 17 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq12KX3xqX1FtSTbFe:-NoqDts520punMXAbBEn:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs#L17)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

Many thanks for adding this fixture!

Please rename to `FullModeDonorImportTests`

and add one more test which is opposite to the existing one, where you assert donors are successfully written to the db when the setting is explicitly set to `true`.
</blockquote></details>

Done.
___
*[`Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs` line 42 at r5](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160#-Noq07X89a12j9otBMEq:-NoqE-uw5Lg4qQpxb_XC:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/5570cb341bb52307e25166e7f4d8045e7226a667/Atlas.DonorImport.Test.Integration/IntegrationTests/Import/RejectFullModeDonorImportTests.cs#L42)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

rename suggestion: `ImportDonorFile_WhenUpdateModeIsFull_AndFullModeImportIsAllowed_DoesNotWriteToDatabase`
</blockquote></details>

Done.

**Review by zabeen — COMMENTED** [review]
> Reviewed 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed, 1 unresolved discussion (waiting on @SergeyEzh)

**Review by zabeen — COMMENTED** [review]
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @SergeyEzh)

**Review by zabeen — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: all files reviewed, 1 unresolved discussion (waiting on @SergeyEzh)

**Review by zabeen — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1160)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @SergeyEzh)

**Pull request merged**